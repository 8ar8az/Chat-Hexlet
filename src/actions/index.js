import { createAction } from 'redux-actions';

export const addChannelSuccess = createAction('CHANNEL_ADD_SUCCESS');

export const updateChannelSuccess = createAction('CHANNEL_UPDATE_SUCCESS');

export const removeChannelRequest = createAction('CHANNEL_REMOVE_REQUEST');
export const removeChannelSuccess = createAction('CHANNEL_REMOVE_SUCCESS');
export const removeChannelFailure = createAction('CHANNEL_REMOVE_FAILURE');

export const changeCurrentChannel = createAction('CURRENT_CHANNEL_CHANGE');

export const addMessagelSuccess = createAction('MESSAGE_ADD_SUCCESS');
