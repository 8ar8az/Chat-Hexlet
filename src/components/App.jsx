import React from 'react';
import { Row } from 'react-bootstrap';

import ChannelsList from './ChannelsList';
import ChannelView from './ChannelView';

const App = () => (
  <Row as="main">
    <ChannelsList />
    <ChannelView />
  </Row>
);

export default App;
