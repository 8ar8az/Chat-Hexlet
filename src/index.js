import '@babel/polyfill';
import 'bootstrap/dist/css/bootstrap.min.css';
import '../assets/application.css';

import gon from 'gon';

import initApp from './app';

if (process.env.NODE_ENV !== 'production') {
  localStorage.debug = 'chat:*';
}

initApp(gon);
