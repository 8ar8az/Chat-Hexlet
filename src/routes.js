import path from 'path';

const API_PART_URL = 'api/v1';
const CHANNELS_PART_URL = 'channels';
const MESSAGES_PART_URL = 'messages';

export default {
  allChannels: () => path.resolve(API_PART_URL, CHANNELS_PART_URL),
  particularChannel: channelId => path.resolve(API_PART_URL, CHANNELS_PART_URL, String(channelId)),
  messagesForParticularChannel: channelId => path.resolve(
    API_PART_URL,
    CHANNELS_PART_URL,
    String(channelId),
    MESSAGES_PART_URL,
  ),
};
