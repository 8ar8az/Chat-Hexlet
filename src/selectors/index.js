/* eslint-disable import/prefer-default-export */
import { createSelector } from 'reselect';
import _ from 'lodash';

const getChannelsById = state => state.channels.byId;
const getChannelsAllIds = state => state.channels.allIds;

export const channelsSelector = createSelector(
  [getChannelsById, getChannelsAllIds],
  (channelsById, channelsAllIds) => {
    const channels = _.map(channelsAllIds, id => channelsById[id]);

    const sortingFn = (channel1, channel2) => {
      if (!channel1.removable && channel2.removable) return -1;
      if (channel1.removable && !channel2.removable) return 1;
      if (_.toUpper(channel1.name) < _.toUpper(channel2.name)) return -1;
      return 1;
    };

    return channels.sort(sortingFn);
  },
);
