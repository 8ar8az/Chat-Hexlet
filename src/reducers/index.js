import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import _ from 'lodash';
import * as actions from '../actions';

const channels = handleActions({
  [actions.addChannelSuccess](state, { payload: { channel } }) {
    const { byId, allIds } = state;
    return {
      ...state,
      byId: { ...byId, [channel.id]: channel },
      allIds: [...allIds, channel.id],
    };
  },
  [actions.removeChannelSuccess](state, { payload: { id } }) {
    const { byId, allIds, currentChannelId } = state;
    return {
      byId: _.omit(byId, id),
      allIds: _.without(allIds, id),
      currentChannelId: currentChannelId === id ? _.find(byId, { name: 'general' }) : currentChannelId,
    };
  },
  [actions.updateChannelSuccess](state, { payload: { channel } }) {
    const { byId } = state;
    return {
      ...state,
      byId: { ...byId, [channel.id]: channel },
    };
  },
  [actions.changeCurrentChannel](state, { payload: { id } }) {
    return {
      ...state,
      currentChannelId: id,
    };
  },
}, { byId: {}, allIds: [], currentChannelId: null });

const channelRemovingState = handleActions({
  [actions.removeChannelRequest](state, { payload: { id } }) {
    return { ...state, [id]: 'requested' };
  },
  [actions.removeChannelSuccess](state, { payload: { id } }) {
    return _.omit(state, id);
  },
  [actions.removeChannelFailure](state, { payload: { id } }) {
    return { ...state, [id]: 'failed' };
  },
}, {});

export default combineReducers({
  channels,
  channelRemovingState,
});
