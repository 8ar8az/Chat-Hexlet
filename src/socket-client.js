import * as actions from './actions';

export default (socketClient, storeDispatch, currentUser, socketLogger) => {
  const logNewMessageFromServer = (messageType, data) => {
    socketLogger('New message from server has been received. Type: "%s". Data: %O', messageType, data);
  };

  const socket = socketClient();

  socket
    .on('connect', () => {
      socketLogger('Socket-client has been connected');
    })
    .on('connect_error', (err) => {
      socketLogger('Error occurs during socket client attempts to connect with the server: %O', err);
    })
    .on('connect_timeout', (timeout) => {
      socketLogger('Timeout occurs during socket client attempts to connect with the server: %s', timeout);
    })
    .on('error', (err) => {
      socketLogger('Error occurs during socket client-server data transfer: %O', err);
    })
    .on('disconnect', (reason) => {
      socketLogger("Socket's client-server connection has been disconnected by reason: %O", reason);
    })
    .on('newChannel', (data) => {
      logNewMessageFromServer('newChannel', data);

      const { data: { attributes: channel } } = data;
      storeDispatch(actions.addChannelSuccess({ channel }));
    })
    .on('removeChannel', (data) => {
      logNewMessageFromServer('removeChannel', data);

      const { data: { id } } = data;
      storeDispatch(actions.removeChannelSuccess({ id }));
    })
    .on('renameChannel', (data) => {
      logNewMessageFromServer('renameChannel', data);

      const { data: { attributes: channel } } = data;
      storeDispatch(actions.updateChannelSuccess({ channel }));
    })
    .on('newMessage', (data) => {
      logNewMessageFromServer('newMessage', data);

      const { data: { attributes: message } } = data;
      if (message.author.id !== currentUser.id) {
        storeDispatch(actions.addMessageSuccess({ message }));
      }
    });
};
