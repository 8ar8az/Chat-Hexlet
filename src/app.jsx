import React from 'react';
import ReactDOM from 'react-dom';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import _ from 'lodash';

import App from './components/App';
import reducers from './reducers';

export const getInitState = ({ channels, messages, currentChannelId }) => {
  const normalizeEntitiesData = (entities) => {
    const allIds = _.map(entities, 'id');
    const byId = _.reduce(entities, (acc, entity) => ({ ...acc, [entity.id]: entity }), {});

    return { byId, allIds };
  };

  return {
    channels: { ...normalizeEntitiesData(channels), currentChannelId },
    messages: normalizeEntitiesData(messages),
  };
};

export default (gon) => {
  /* eslint-disable no-underscore-dangle */
  const devTools = window !== 'undefined'
    && window.__REDUX_DEVTOOLS_EXTENSION__
    && window.__REDUX_DEVTOOLS_EXTENSION__();
  /* eslint-enable no-underscore-dangle */

  const initState = getInitState(gon);
  const store = createStore(
    reducers,
    initState,
    devTools,
  );

  ReactDOM.render(
    <Provider store={store}>
      <App />
    </Provider>,
    document.getElementById('chat'),
  );
};
