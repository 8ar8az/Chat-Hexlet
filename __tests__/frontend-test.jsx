import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import App from '../src/components/App';
import reducers from '../src/reducers';
import { getInitState } from '../src/app';
import { defaultState } from '../server/routes';

Enzyme.configure({ adapter: new Adapter() });

let wrapper;

beforeAll(() => {
  const store = createStore(
    reducers,
    getInitState(defaultState),
  );

  const vdom = (
    <Provider store={store}>
      <App />
    </Provider>
  );

  wrapper = mount(vdom);
});

test('Init state render', () => {
  expect(wrapper.render()).toMatchSnapshot();
});
