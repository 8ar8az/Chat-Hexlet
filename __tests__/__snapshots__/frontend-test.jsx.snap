// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Init state render 1`] = `
<main
  class="row"
>
  <aside
    class="col-4"
  >
    <ul
      class="list-group"
    >
      <li
        class="d-flex align-items-center list-group-item active list-group-item-action"
      >
        general
        <div
          class="ml-auto btn-group btn-group-sm"
          role="group"
        >
          <button
            class="btn btn-success"
            type="button"
          >
            Edit
          </button>
        </div>
      </li>
      <li
        class="d-flex align-items-center list-group-item list-group-item-action"
      >
        random
        <div
          class="ml-auto btn-group btn-group-sm"
          role="group"
        >
          <button
            class="btn btn-success"
            type="button"
          >
            Edit
          </button>
        </div>
      </li>
      <li
        class="d-flex align-items-center list-group-item list-group-item-action"
      >
        my-channel
        <div
          class="ml-auto btn-group btn-group-sm"
          role="group"
        >
          <button
            class="btn btn-success"
            type="button"
          >
            Edit
          </button>
          <button
            class="btn btn-danger"
            type="button"
          >
            Delete
          </button>
        </div>
      </li>
    </ul>
  </aside>
  <section
    class="col"
  />
</main>
`;
